tests/data_context/cloud_data_context/test_datasource_crud.py

```diff
@@ -134,3 +136,110 @@ def json(self):
+
+
+@pytest.mark.cloud
+@pytest.mark.integration
+@pytest.mark.parametrize(
+    "config_includes_name_setting",
+    [
+        pytest.param("name_supplied_separately", id="name supplied separately"),
+        pytest.param("config_includes_name", id="config includes name"),
+        pytest.param(
+            "name_supplied_separately_and_included_in_config",
+            id="name supplied separately and config includes name",
+            marks=pytest.mark.xfail(strict=True, raises=TypeError),
+        ),
+    ],
+)
+def test_data_context_in_cloud_mode_add_datasource(
+    config_includes_name_setting: str,
+    empty_data_context_in_cloud_mode: DataContext,
+    datasource_config: DatasourceConfig,
+    datasource_name: str,
+    ge_cloud_base_url: str,
+    ge_cloud_organization_id: str,
+    request_headers: dict,
+):
+    """A DataContext in cloud mode should save to the cloud backed Datasource store when calling add_datasource. When saving, it should use the id from the response
+    to create the datasource."""
+
+    context: DataContext = empty_data_context_in_cloud_mode
+    # Make sure we are using the right fixture
+    assert isinstance(context, DataContext)
+    assert isinstance(
+        context._datasource_store._store_backend,
+        GeCloudStoreBackend,
+    )
+    assert context.ge_cloud_mode
+    assert len(context.list_datasources()) == 0
+
+    # Setup
+    datasource_id: str = "some_uuid"
+    datasource_config_with_name: DatasourceConfig = copy.deepcopy(datasource_config)
+    datasource_config_with_name.name = datasource_name
+
+    def mocked_post_response(*args, **kwargs):
+        class MockResponse:
+            def __init__(self, json_data: dict, status_code: int) -> None:
+                self._json_data = json_data
+                self._status_code = status_code
+
+            def json(self):
+                return self._json_data
+
+        return MockResponse({"data": {"id": datasource_id}}, 201)
+
+    with patch(
+        "requests.post", autospec=True, side_effect=mocked_post_response
+    ) as mock_post:
+
+        # Call add_datasource with and without the name field included in the datasource config
+        stored_datasource: BaseDatasource
+        if config_includes_name_setting == "name_supplied_separately":
+            stored_datasource = context.add_datasource(
+                name=datasource_name,
+                **datasource_config.to_dict(),
+            )
+        elif config_includes_name_setting == "config_includes_name":
+            stored_datasource = context.add_datasource(
+                **datasource_config_with_name.to_dict()
+            )
+        elif (
+            config_includes_name_setting
+            == "name_supplied_separately_and_included_in_config"
+        ):
+            stored_datasource = context.add_datasource(
+                name=datasource_name,
+                **datasource_config_with_name.to_dict(),
+            )
+
```

Comment:
with all the branching here, should we split this out of a parameterized test?

Reply:
For sure - I promised to make that change in the last PR will make it here or in the next.